forte/data/data_pack_dataset.py

```diff
@@ -0,0 +1,278 @@
+    :class:`forte.data.data_pack_dataset.DataPackIterator`.
+
+    Args:
+        pack_generator (Iterator[DataPack]): A generator of
+            :class:`forte.data.data_pack.DataPack`.
+        context_type: The granularity of a single example which
+            could be any ``Annotation`` type. For example, it can be
+            :class:`ft.onto.base_ontology.Sentence`, then each training example
+            will represent the information of a sentence.
+        request: The request of type `Dict` sent to
+            :class:`forte.data.readers.base_reader.PackReader` to query
+            specific data.
+        skip_k (int): Will skip the first `skip_k` instances and generate
+            data from the (`offset` + 1)th instance.
+
+    .. note::
+        For parameters `context_type`, `request`, `skip_k`, please refer to
+        :meth:`get_data()` in :class:`forte.data.data_pack.DataPack`.
+    """
+    def __init__(self,
+                 pack_generator: Iterator[DataPack],
+                 context_type: Type[Annotation],
+                 request: Optional[DataRequest] = None,
+                 skip_k: int = 0):
+        self._iterator: Iterator = DataPackIterator(pack_generator,
+                                                    context_type,
+                                                    request,
+                                                    skip_k)
+        super().__init__(self)
+
+    def __iter__(self):
+        return self._iterator
+
+
+class DataPackDataset(DatasetBase):
+    """
+    A dataset representing data packs. Calling an
+    `DataIterator
+    <https://texar-pytorch.readthedocs.io/en/latest/code/data.html#dataiterator>`_
+    over this `DataPackDataset` will produce an `Iterate` over batch of examples
+    parsed by a reader from given data packs.
+
+    Args:
+        data_source: A data source of type
+            :class:`forte.data.data_pack_dataset.DataPackDataSource`.
+        feature_schemes (dict): A `Dict` containing all the information to do
+            data pre-processing. This is exactly the same as the `schemes` in
+            `feature_resource`. Please refer to :meth:`feature_resource` in
+            :class:`forte.train_preprocessor.TrainPreprocessor` for details.
+        hparams: A `dict` or instance of :
+            class:`forte.common.configuration.Config` containing
+            hyperparameters. See :meth:`default_hparams` in
+            `DatasetBase
+            <https://texar-pytorch.readthedocs.io/en/latest/code/data.html#datasetbase>`
+            for the defaults.
+        device: The device of the produced batches. For GPU training,
+            set to current CUDA device.
+    """
+    def __init__(self,
+                 data_source: DataPackDataSource,
+                 feature_schemes: Dict,
+                 hparams: Union[Dict, HParams] = None,
+                 device: Optional[torch.device] = None):
+        self._data_source = data_source
```

Comment:
I think adding type for class variables would help the most.

Reply:
Added.